File: lib/patches/unmask_repos.rb
# frozen_string_literal: true

# SPDX-FileCopyrightText: Copyright (c) 2024-2026 Zerocracy
# SPDX-License-Identifier: MIT

require 'fbe/unmask_repos'

module Fbe
  class << self
    def unmask_repos( # rubocop:disable Elegant/GoodMethodName
      options: $options, global: $global, loog: $loog, epoch: $epoch || Time.now, kickoff: $kickoff || Time.now,
      quota_aware: true, lifetime_aware: true, timeout_aware: true
    )
      raise(Fbe::Error, 'Repositories mask is not specified') unless options.repositories
      raise(Fbe::Error, 'Repositories mask is empty') if options.repositories.empty?
      return if block_given? && Fbe.over?(
        global:, options:, loog:, epoch:, kickoff:, quota_aware:, lifetime_aware:, timeout_aware:
      )
      repos = []
      octo = Fbe.octo(loog:, global:, options:)
      masks = (options.repositories || '').split(',')
      masks.reject { |m| m.start_with?('-') }.each do |mask|
        unless mask.include?('*')
          repos << mask
          next
        end
        re = Fbe.mask_to_regex(mask)
        org = mask.split('/')[0]
        list =
          if org == '*'
            octo.repositories
          else
            begin
              octo.organization_repositories(org, type: 'all')
            rescue Octokit::NotFound, Octokit::Forbidden, Octokit::Unauthorized,
              Octokit::ServerError, Octokit::TooManyRequests => e
              $loog.warn("[#{$judge}] Failed to list repos for org #{org}: #{e.class}: #{e.message}")
              octo.repositories(org)
            rescue Net::OpenTimeout, Net::ReadTimeout, SocketError, Errno::ECONNRESET => e
              $loog.warn("[#{$judge}] Network error listing repos for org #{org}: #{e.message}")
              octo.repositories(org)
            end
          end
        list.each do |r|
          repos << r[:full_name] if re.match?(r[:full_name])
        end
      end
      masks.select { |m| m.start_with?('-') }.each do |mask|
        re = Fbe.mask_to_regex(mask[1..])
        repos.reject! { |r| re.match?(r) }
      end
      repos.reject! do |repo|
        octo.repository(repo)[:archived]
      rescue Octokit::NotFound, Octokit::Deprecated => e
        $loog.info("Repository #{repo} not found: #{e.message}")
        false
      rescue Octokit::Forbidden, Octokit::Unauthorized => e
        $loog.warn(
          "[#{$judge}] Access forbidden to #{repo} " \
          "(transient, will retry next cycle): #{e.class}: #{e.message}"
        )
        false
      rescue Octokit::TooManyRequests, Octokit::ServerError,
        Net::OpenTimeout, Net::ReadTimeout, SocketError, Errno::ECONNRESET => e
        $loog.warn("[#{$judge}] Transient error checking #{repo}: #{e.class}: #{e.message}")
        false
      end
      raise(Fbe::Error, "No repos found matching: #{options.repositories.inspect}") if repos.empty?
      repos.shuffle!
      loog.debug("Scanning #{repos.size} repositories: #{repos.joined}...")
      repos.each do |repo|
        octo.repository(repo)
      rescue Octokit::NotFound, Octokit::Deprecated => e
        $loog.debug("[#{$judge}] Repository #{repo} not found: #{e.message}")
      rescue Octokit::Forbidden, Octokit::Unauthorized, Octokit::TooManyRequests => e
        $loog.warn("[#{$judge}] Repository #{repo} temporarily unavailable: #{e.class}: #{e.message}")
      rescue Net::OpenTimeout, Net::ReadTimeout, SocketError, Errno::ECONNRESET => e
        $loog.warn("[#{$judge}] Network error checking #{repo}: #{e.message}")
      end
      return repos unless block_given?
      repos.each do |repo|
        break if Fbe.over?(global:, options:, loog:, epoch:, kickoff:, quota_aware:, lifetime_aware:, timeout_aware:)
        yield(repo)
      end
    end
  end
end

File: lib/patches/unmask_repos.py
# complete code
import logging

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class OctokitError(Exception):
    """Base class for Octokit errors"""
    pass

class NotFoundError(OctokitError):
    """Raised when a resource is not found"""
    pass

class DeprecatedError(OctokitError):
    """Raised when an endpoint is deprecated"""
    pass

class ForbiddenError(OctokitError):
    """Raised when access is forbidden"""
    pass

def unmask_repos(fbe, fact):
    """
    Unmask repositories for a given fact.

    Args:
        fbe (Fbe): Fbe instance
        fact (dict): Fact dictionary

    Returns:
        list: List of unmasked repositories
    """
    logger.info("Unmasking repositories for fact %s", fact)
    try:
        # Get repository workflow runs
        workflow_runs = fbe.octo.repository_workflow_runs(fact['repo'], created=fact['created'])
        if workflow_runs is None:
            raise NotFoundError("Repository workflow runs not found")
        elif isinstance(workflow_runs, list) and len(workflow_runs) == 0:
            raise NotFoundError("No repository workflow runs found")
        elif isinstance(workflow_runs, dict) and 'message' in workflow_runs and workflow_runs['message'] == 'Not Found':
            raise NotFoundError("Repository workflow runs not found")

        # Get workflow run usage
        usage = fbe.octo.workflow_run_usage(fact['repo'], workflow_runs[0]['id'])
        if usage is None:
            raise NotFoundError("Workflow run usage not found")
        elif isinstance(usage, dict) and 'message' in usage and usage['message'] == 'Not Found':
            raise NotFoundError("Workflow run usage not found")

        # Return unmasked repositories
        return [repo for repo in fact['repos'] if repo['id'] in [run['id'] for run in workflow_runs]]

    except NotFoundError as e:
        logger.info("Repository not found: %s", e)
        return []
    except DeprecatedError as e:
        logger.info("Deprecated endpoint: %s", e)
        return []
    except ForbiddenError as e:
        logger.info("Forbidden access: %s", e)
        logger.warning("Will retry next cycle")
        return []
    except Exception as e:
        logger.error("Error unmasking repositories: %s", e)
        return []